README
TODO description

Install
=======

TODO python python3

TODO pip TODO virtualenv

  python3 setup.py install

Documentation
=============

You can build the documentation using:

  cd doc && make html


CHANGELOG
Version 0.1.2 (2015/03/14)

  - New version to solve packaging issues.

Version 0.1.1 (2015/03/14)

  - Bugs
    - Closes bug: Paper size was not case sensitive

  - Meta
    - Changed project URL to https://git.framasoft.org/spalax/papersize

  - Internal
    - Internal change: module is no longer imported to run setup.py or build
      documentation.
    - Added tests

Version 0.1.0 (2015/01/24)

  - First published version.


papersize/test.py
# Copyright 2015 Louis Paternault
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""Tests"""

from decimal import Decimal
import doctest
import os
import unittest

import papersize

class TestParse(unittest.TestCase):
    """Test parsing related functions."""
    # pylint: disable = invalid-name, star-args

    def assertIterAlmostEqual(self, iter1, iter2):
        """Assert iterators of elements are almost items.

        Both arguments are expected to be iterators of the same size, and
        iterators with the same indexes are checked to be almost equal.
        """


        for left, right in zip(iter1, iter2):
            self.assertAlmostEqual(Decimal(left), Decimal(right))

    def testParseLength(self):
        """Test :func:`papersize.parse_length`."""
        for (args, result) in [
                (("10cm", "mm"), 100),
                (("10in",), 722.7),
            ]:
            self.assertAlmostEqual(
                papersize.parse_length(*args),
                Decimal(result),
                )

    def testParseCouple(self):
        """Test :func:`papersize.parse_couple`."""
        for (args, result) in [
                (("10cmx1mm",), (284.5275591, 2.845275591)),
                (("10cmx1mm", "mm"), (100, 1)),
                (("2pc x 3dd", "pt"), (24, 3.21)),
            ]:
            self.assertIterAlmostEqual(
                papersize.parse_couple(*args),
                result,
                )

    def testParsePaperSize(self):
        """Test :func:`papersize.parse_papersize`."""
        for (args, result) in [
                (("a4", "cm"), (21, 29.7)),
                (("20cm x 1mm", "cm"), (20, 0.1)),
                (("quarto", ), (650.43, 794.97)),
            ]:
            self.assertIterAlmostEqual(
                papersize.parse_papersize(*args),
                result,
                )

    def testConvertLength(self):
        """Test :func:`papersize.convert_length`."""
        for (args, result) in [
                ((10, "cm", "mm"), 100),
                ((1, "mm", "pt"), 2.845275591),
            ]:
            self.assertAlmostEqual(
                papersize.convert_length(*args),
                Decimal(result),
                )

class TestOrientation(unittest.TestCase):
    """Test orientation related tools."""
    # pylint: disable = invalid-name, star-args

    def testPortraitLandscape(self):
        """Test portrait/landscape functions."""
        self.assertTrue(papersize.is_portrait(10, 11))
        self.assertTrue(papersize.is_portrait(10, 10))
        self.assertFalse(papersize.is_portrait(11, 10))

        self.assertFalse(papersize.is_landscape(10, 11))
        self.assertTrue(papersize.is_landscape(10, 10))
        self.assertTrue(papersize.is_landscape(11, 10))

        self.assertFalse(papersize.is_square(10, 11))
        self.assertTrue(papersize.is_square(10, 10))
        self.assertFalse(papersize.is_square(11, 10))

    def testRotate(self):
        """Test :func:`papersize.rotate` function."""
        self.assertEqual(
            papersize.rotate((10, 11), True),
            (10, 11),
            )
        self.assertEqual(
            papersize.rotate((10, 11), False),
            (11, 10),
            )

def suite():
    """Return a :class:`TestSuite` object, testing all module :mod:`papersize`.
    """
    test_loader = unittest.defaultTestLoader
    return test_loader.discover(os.path.dirname(__file__))

def load_tests(__loader, tests, __pattern):
    """Load tests (unittests and doctests)."""
    # Loading doctests
    tests.addTests(doctest.DocTestSuite(papersize))

    # Unittests are loaded by default

    return tests

if __name__ == "__main__":
    unittest.TextTestRunner().run(suite())


papersize/main.py
# Copyright Louis Paternault 2011-2014
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>. 1

"""Paper-size related functions"""

import decimal
import os
import re
import subprocess

from pdfautonup import errors

PAPERSIZES = {
    # http://www.printernational.org/iso-paper-sizes.php
    "4a0": "1682mm x 2378mm",
    "2a0": "1189mm x 1682mm",
    "a0": "841mm x 1189mm",
    "a1": "594mm x 841mm",
    "a2": "420mm x 594mm",
    "a3": "297mm x 420mm",
    "a4": "210mm x 297mm",
    "a5": "148mm x 210mm",
    "a6": "105mm x 148mm",
    "a7": "74mm x 105mm",
    "a8": "52mm x 74mm",
    "a9": "37mm x 52mm",
    "a10": "26mm x 37mm",

    "b0": "1000mm x 1414mm",
    "b1": "707mm x 1000mm",
    "b2": "500mm x 707mm",
    "b3": "353mm x 500mm",
    "b4": "250mm x 352mm",
    "b5": "176mm x 250mm",
    "b6": "125mm x 176mm",
    "b7": "88mm x 125mm",
    "b8": "62mm x 88mm",
    "b9": "44mm x 62mm",
    "b10": "31mm x 44mm",

    # http://www.paper-sizes.com/north-american-paper-sizes/north-american-architectural-paper-sizes
    "archA": "9in x 12in",
    "archB": "12in x 18in",
    "archC": "18in x 24in",
    "archD": "24in x 36in",
    "archE": "36in x 48in",

    # http://www.engineeringtoolbox.com/office-paper-sizes-d_213.html
    "letter": "8.5in x 11in",
    "legal": "8.5in x 14in",
    "executive": "7in x 10in",
    "tabloid": "11in x 17in",
    "statement": "5in x 8.5in",
    "halfletter": "5in x 8.5in",
    "folio": "8in x 13in",

    # http://www.paper-sizes.com/north-american-paper-sizes/north-american-loose-paper-sizes
    "ledger": "17in x 11in",

    # http://simple.wikipedia.org/wiki/Paper_size
    "quarto": "8in x 10in",

    # http://hplipopensource.com/hplip-web/tech_docs/page_sizes.html
    "flsa": "8.5in x 13in",

    # http://www.coding-guidelines.com/numbers/ndb/units/area.txt
    "flse": "8.5in x 13in",

    # http://jexcelapi.sourceforge.net/resources/javadocs/2_6_10/docs/jxl/format/PaperSize.html
    "note": "8.5in x 11in",
    "11x17": "11in x 17in",
    "10x14": "10in x 14in",

    }

# Source: http://en.wikibooks.org/wiki/LaTeX/Lengths
UNITS = {
    "": 1, # Default is point (pt)
    "pt": 1,
    "mm": 2.84,
    "cm": 28.4,
    "in": 72.27,
    "bp": 1.00375,
    "pc": 12,
    "dd": 1.07,
    "cc": 12.84,
    "sp": 0.000015,
    }

UNITS_RE = r"({})".format("|".join(UNITS.keys()))
SIZE_RE = r"([\d.]+){}".format(UNITS_RE)
PAPERSIZE_RE = r"^(?P<width>{size}) *[x× ]? *(?P<height>{size})$".format(
    size=SIZE_RE
    )

UNITS_COMPILED_RE = re.compile("^{}$".format(UNITS_RE))
SIZE_COMPILED_RE = re.compile("^{}$".format(SIZE_RE).format("size"))
PAPERSIZE_COMPILED_RE = re.compile(PAPERSIZE_RE.format("width", "height"))

def parse_length(string):
    """Return a length corresponding to the string.

    :rtype: :class:`decimal.Decimal`
    :return: The length, in points.
    """
    match = SIZE_COMPILED_RE.match(string).groups()
    return decimal.Decimal(match[0]) * decimal.Decimal(UNITS[match[1]])

def parse_couple(string):
    """Return a tuple of dimensions.

    :param str string: The string to parse, as "LENGTHxLENGTH" (where LENGTH
    are length).
    :return: A tuple of :class:`decimal.Decimal`, reprenting the dimensions,
        in points.
    """
    try:
        match = PAPERSIZE_COMPILED_RE.match(string).groupdict()
        return (
            parse_length(match['width']),
            parse_length(match['height']),
            )
    except AttributeError:
        raise errors.CouldNotParse(string)

def parse_paper_size(string):
    """Return the papersize corresponding to string.
    """
    if string.lower() in PAPERSIZES:
        return parse_paper_size(PAPERSIZES[string])
    return parse_couple(string)

def parse_lc_paper(string):
    """Parse LC_PAPER locale variable

    We assume units are milimeters.
    """
    dimensions = {}
    for line in string.split('\n'):
        if line.startswith("width="):
            dimensions["width"] = parse_length("{}mm".format(line[6:]))
        if line.startswith("height="):
            dimensions["height"] = parse_length("{}mm".format(line[7:]))
    if len(dimensions) == 2:
        return (dimensions["width"], dimensions["height"])
    raise errors.CouldNotParse(string)

def target_paper_size(target_size):
    """Return the target paper size.

    :param str target_size: Target size, if provided by user in command line.
    """
    # pylint: disable=too-many-return-statements

    # Option set by user on command line
    if target_size is not None:
        return parse_paper_size(target_size[0])

    # LC_PAPER environment variable (can be read from "locale -k LC_PAPER"
    try:
        return parse_lc_paper(subprocess.check_output(
            ["locale", "-k", "LC_PAPER"],
            universal_newlines=True,
            ))
    except (subprocess.CalledProcessError, errors.CouldNotParse):
        pass

    # PAPERSIZE environment variable
    try:
        return parse_paper_size(os.environ['PAPERSIZE'].strip())
    except KeyError:
        pass

    # file described by the PAPERCONF environment variable
    try:
        return parse_paper_size(
            open(os.environ['PAPERCONF'], 'r').read().strip()
            )
    except errors.CouldNotParse:
        raise
    except: # pylint: disable=bare-except
        pass

    # content of /etc/papersize
    try:
        return parse_paper_size(open('/etc/papersize', 'r').read().strip())
    except errors.CouldNotParse:
        raise
    except: # pylint: disable=bare-except
        pass

    # stdout of the paperconf command
    try:
        return parse_paper_size(subprocess.check_output(
            ["paperconf"],
            universal_newlines=True,
            ).strip())
    except (subprocess.CalledProcessError, errors.CouldNotParse):
        pass

    # Eventually, if everything else has failed, a4
    return parse_paper_size('a4')


papersize/__init__.py
#!/usr/bin python
# -*- coding: utf8 -*-

# Copyright Louis Paternault 2015
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>. 1

"""Paper size related data and functions

In this module:

- the default unit (input and output) is point (``pt``);
- every numbers are returned as :class:`decimal.Decimal` objects.


Constants
---------

.. autodata:: UNITS
    :annotation:

.. autodata:: SIZES
    :annotation:

.. autodata:: PORTRAIT
    :annotation:

.. autodata:: LANDSCAPE
    :annotation:

Unit conversion
---------------

.. autofunction:: convert_length

Parsers
-------

.. autofunction:: parse_length

.. autofunction:: parse_couple

.. autofunction:: parse_papersize

Paper orientation
-----------------

.. autofunction:: is_portrait

.. autofunction:: is_landscape

.. autofunction:: is_square

.. autofunction:: rotate

Exceptions
----------

.. autoclass:: PapersizeException

.. autoclass:: CouldNotParse

.. autoclass:: UnknownOrientation

"""

from __future__ import unicode_literals
from decimal import Decimal
import os
import re

__version__ = "0.1.1"
__AUTHOR__ = "Louis Paternault (spalax@gresille.org)"
__COPYRIGHT__ = "(C) 2015 Louis Paternault. GNU GPL 3 or later."

SIZES = {
    # http://www.printernational.org/iso-paper-sizes.php
    "4a0": "1682mm x 2378mm",
    "2a0": "1189mm x 1682mm",
    "a0": "841mm x 1189mm",
    "a1": "594mm x 841mm",
    "a2": "420mm x 594mm",
    "a3": "297mm x 420mm",
    "a4": "210mm x 297mm",
    "a5": "148mm x 210mm",
    "a6": "105mm x 148mm",
    "a7": "74mm x 105mm",
    "a8": "52mm x 74mm",
    "a9": "37mm x 52mm",
    "a10": "26mm x 37mm",

    "b0": "1000mm x 1414mm",
    "b1": "707mm x 1000mm",
    "b2": "500mm x 707mm",
    "b3": "353mm x 500mm",
    "b4": "250mm x 352mm",
    "b5": "176mm x 250mm",
    "b6": "125mm x 176mm",
    "b7": "88mm x 125mm",
    "b8": "62mm x 88mm",
    "b9": "44mm x 62mm",
    "b10": "31mm x 44mm",

    # http://www.paper-sizes.com/north-american-paper-sizes/north-american-architectural-paper-sizes
    "archA": "9in x 12in",
    "archB": "12in x 18in",
    "archC": "18in x 24in",
    "archD": "24in x 36in",
    "archE": "36in x 48in",

    # http://www.engineeringtoolbox.com/office-paper-sizes-d_213.html
    "letter": "8.5in x 11in",
    "legal": "8.5in x 14in",
    "executive": "7in x 10in",
    "tabloid": "11in x 17in",
    "statement": "5in x 8.5in",
    "halfletter": "5in x 8.5in",
    "folio": "8in x 13in",

    # http://www.paper-sizes.com/north-american-paper-sizes/north-american-loose-paper-sizes
    "ledger": "17in x 11in",

    # http://simple.wikipedia.org/wiki/Paper_size
    "quarto": "9in x 11in",

    # http://hplipopensource.com/hplip-web/tech_docs/page_sizes.html
    "flsa": "8.5in x 13in",

    # http://www.coding-guidelines.com/numbers/ndb/units/area.txt
    "flse": "8.5in x 13in",

    # http://jexcelapi.sourceforge.net/resources/javadocs/2_6_10/docs/jxl/format/PaperSize.html
    "note": "8.5in x 11in",
    "11x17": "11in x 17in",
    "10x14": "10in x 14in",

    }
"""Dictionary of named sizes.

Keys are names (e.g. ``a4``, ``letter``) and values are strings,
human-readable, and parsable by :func:`parse_papersize` (e.g. ``21cm x
29.7cm``).
"""

# Source: http://en.wikibooks.org/wiki/LaTeX/Lengths
_TXT_UNITS = {
    "": "1", # Default is point (pt)
    "pt": "1",
    "mm": "2.845275591",
    "cm": "28.45275591",
    "in": "72.27",
    "bp": "1.00375",
    "pc": "12",
    "dd": "1.07",
    "cc": "12.84",
    "sp": "0.000015",
    }

UNITS = dict([
    (key, Decimal(value))
    for (key, value)
    in _TXT_UNITS.items()
    ])
"""Dictionary of units.

Keys are unit abbreviation (e.g. ``pt`` or ``cm``), and values are their value
in points (e.g. ``UNITS['pt']`` is 1, ``UNITS['pc']``] is 12), as
:class:`decimal.Decimal` objects.
"""

PORTRAIT = True
"""Constant corresponding to the portrait orientation

That is, height greater than width.
"""

LANDSCAPE = False
"""Constant corresponding to the landscape orientation

That is, width greater than height.
"""

__UNITS_RE = r"({})".format("|".join(UNITS.keys()))
__SIZE_RE = r"([\d.]+){}".format(__UNITS_RE)
__PAPERSIZE_RE = r"^(?P<width>{size}) *[x× ]? *(?P<height>{size})$".format(
    size=__SIZE_RE
    )

__SIZE_COMPILED_RE = re.compile("^{}$".format(__SIZE_RE).format("size"))
__PAPERSIZE_COMPILED_RE = re.compile(__PAPERSIZE_RE.format("width", "height"))

class PapersizeException(Exception):
    """All exceptions of this module inherit from this one."""
    pass

class CouldNotParse(PapersizeException):
    """Raised when a string could not be parsed.

    :param str string: String that could not be parsed.
    """
    def __init__(self, string):
        super(CouldNotParse, self).__init__()
        self.string = string

    def __str__(self):
        return "Could not parse string '{}'.".format(self.string)

class UnknownOrientation(PapersizeException):
    """Raised when a string could not be parsed.

    :param obj string: Object wrongly provided as an orientation.
    """
    def __init__(self, string):
        super(UnknownOrientation, self).__init__()
        self.string = string

    def __str__(self):
        return (
            "'{}' is not one of `papersize.PORTRAIT` or `papersize.LANDSCAPE`"
            ).format(self.string)

def convert_length(length, orig, dest):
    """Convert length from one unit to another.

    :param decimal.Decimal length: Length to convert, as any object convertible
        to a :class:`decimal.Decimal`.
    :param str orig: Unit of ``length``, as a string which is a key of
        :data:`UNITS`.
    :param str dest: Unit in which ``length`` will be converted, as a string
        which is a key of :data:`UNITS`.

    Due to floating point arithmetic, there can be small rounding errors.

    >>> convert_length(0.1, "cm", "mm")
    Decimal('1.000000000000000055511151231')
    """
    return (Decimal(UNITS[orig]) * Decimal(length)) / Decimal(UNITS[dest])

def parse_length(string, unit="pt"):
    """Return a length corresponding to the string.

    :param str string: The string to parse, as a length and a unit, for
        instance ``10.2cm``.
    :param str unit: The unit of the return value, as a key of :data:`UNITS`.
    :return: The length, in an unit given by the ``unit`` argument.
    :rtype: :class:`decimal.Decimal`

    >>> parse_length("1cm", "mm")
    Decimal('1E+1')
    >>> parse_length("1cm", "cm")
    Decimal('1')
    >>> parse_length("10cm")
    Decimal('284.52755910')
    """
    match = __SIZE_COMPILED_RE.match(string)
    if match is None:
        raise CouldNotParse(string)
    return convert_length(
        Decimal(match.groups()[0]),
        match.groups()[1],
        unit,
        )

def parse_couple(string, unit="pt"):
    """Return a tuple of dimensions.

    :param str string: The string to parse, as "LENGTHxLENGTH" (where LENGTH
        are length, parsable by :func:`parse_length`). Example: ``21cm x
        29.7cm``. The separator can be ``x``, ``×`` or empty, surrounded by an
        arbitrary number of spaces. For instance: ``2cmx3cm``, ``2cm x 3cm``,
        ``2cm×3cm``, ``2cm 3cm``.
    :rtype: :class:`tuple`
    :return: A tuple of :class:`decimal.Decimal`, representing the dimensions.

    >>> parse_couple("1cm 10cm", "mm")
    (Decimal('1E+1'), Decimal('1.0E+2'))
    >>> parse_couple("1mm 10mm", "cm")
    (Decimal('0.1'), Decimal('1.0'))
    """
    try:
        match = __PAPERSIZE_COMPILED_RE.match(string).groupdict()
        return (
            parse_length(match['width'], unit),
            parse_length(match['height'], unit),
            )
    except AttributeError:
        raise CouldNotParse(string)

def parse_papersize(string, unit="pt"):
    """Return the papersize corresponding to string.

    :param str string: The string to parse. It can be either a named size (as
        keys of constant :data:`SIZES`), or a couple of lengths (that will be
        processed by :func:`parse_couple`). The named paper sizes are case
        insensitive.  The following strings return the same size: ``a4``,
        ``A4``, ``21cm 29.7cm``, ``210mmx297mm``, ``21cm  ×  297mm``…
    :param str unit: The unit of the return values.
    :return: The paper size, as a couple of :class:`decimal.Decimal`.
    :rtype: :class:`tuple`

    >>> parse_papersize("A4", "cm")
    (Decimal('21.0'), Decimal('29.7'))
    >>> parse_papersize("21cm x 29.7cm", "mm")
    (Decimal('2.1E+2'), Decimal('297'))
    >>> parse_papersize("10 100")
    (Decimal('10'), Decimal('100'))
    """
    if string.lower() in SIZES:
        return parse_papersize(SIZES[string.lower()], unit)
    return parse_couple(string, unit)

def is_portrait(width, height):
    """Return whether paper orientation is portrait

    That is, height greater or equal to width.

    :param width: Width of paper, as any sortable object.
    :param height: Height of paper, as any sortable object.

    >>> is_portrait(11, 10)
    False
    >>> is_portrait(10, 10)
    True
    >>> is_portrait(10, 11)
    True
    """
    return width <= height

def is_landscape(width, height):
    """Return whether paper orientation is landscape

    That is, width greater or equal to height.

    :param width: Width of paper, as any sortable object.
    :param height: Height of paper, as any sortable object.

    >>> is_landscape(11, 10)
    True
    >>> is_landscape(10, 10)
    True
    >>> is_landscape(10, 11)
    False
    """
    return height <= width

def is_square(width, height):
    """Return whether paper is a square (width equals height).

    :param width: Width of paper, as any sortable object.
    :param height: Height of paper, as any sortable object.

    >>> is_square(11, 10)
    False
    >>> is_square(10, 10)
    True
    >>> is_square(10, 11)
    False
    """
    return width == height

def rotate(size, orientation):
    """Return the size, rotated if necessary to make it portrait or landscape.

    :param tuple size: Couple paper of dimension, as sortable objects
        (:class:`int`, :class:`float`, :class:`decimal.Decimal`…).
    :param orientation: Return format, one of ``PORTRAIT`` or ``LANDSCAPE``.
    :return: The size, as a couple of dimensions, of the same type of the
        ``size`` parameter.
    :rtype: :class:`tuple`

    >>> rotate((21, 29.7), PORTRAIT)
    (21, 29.7)
    >>> rotate((21, 29.7), LANDSCAPE)
    (29.7, 21)
    """
    if orientation == PORTRAIT:
        return (min(size), max(size))
    elif orientation == LANDSCAPE:
        return (max(size), min(size))
    else:
        raise UnknownOrientation(orientation)
